src/Nethermind/Nethermind.Evm/IntrinsicGasCalculator.cs

```diff
@@ -7,27 +7,56 @@
 using System.IO;
 using Nethermind.Core;
 using Nethermind.Core.Eip2930;
+using Nethermind.Core.Extensions;
 using Nethermind.Core.Specs;
 using Nethermind.Int256;
-using Nethermind.Core.Extensions;
 
 namespace Nethermind.Evm;
 
+
+public readonly record struct IntrinsicGas(long Standard, long FloorGas)
+{
+    public long MinimalGas { get; } = Math.Max(Standard, FloorGas);
+    public static explicit operator long(IntrinsicGas gas) => gas.MinimalGas;
+}
+
 public static class IntrinsicGasCalculator
 {
-    public static long Calculate(Transaction transaction, IReleaseSpec releaseSpec) =>
-        GasCostOf.Transaction
-        + DataCost(transaction, releaseSpec)
-        + CreateCost(transaction, releaseSpec)
-        + AccessListCost(transaction, releaseSpec)
-        + AuthorizationListCost(transaction, releaseSpec);
+    public static IntrinsicGas Calculate(Transaction transaction, IReleaseSpec releaseSpec)
+    {
+        var intrinsicGas = GasCostOf.Transaction
+               + DataCost(transaction, releaseSpec)
+               + CreateCost(transaction, releaseSpec)
+               + AccessListCost(transaction, releaseSpec)
+               + AuthorizationListCost(transaction, releaseSpec);
+        var floorGas = CalculateFloorCost(transaction, releaseSpec);
+        return new IntrinsicGas(intrinsicGas, floorGas);
+    }
+
+    public static long CalculateFloorCost(Transaction transaction, IReleaseSpec releaseSpec)
+    {
+        if (!releaseSpec.IsEip7623Enabled) return 0;
+
+        long txDataNonZeroMultiplier = releaseSpec.IsEip2028Enabled
+            ? GasCostOf.TxDataNonZeroMultiplierEip2028
+            : GasCostOf.TxDataNonZeroMultiplier;
+        Span<byte> data = transaction.Data.GetValueOrDefault().Span;
+
+        int totalZeros = data.CountZeros();
+
+        var tokensInCallData = totalZeros + (data.Length - totalZeros) * txDataNonZeroMultiplier;
+
+        return GasCostOf.Transaction + tokensInCallData * GasCostOf.TotalCostFloorPerTokenEip7623;
```

Comment:
This calculation in `CalculateFloorCost` has very similar code to the one in `DataCost`, this could be unified in code.

Reply:
https://github.com/NethermindEth/nethermind/pull/7859#discussion_r1870933575
i maybe understood your comment wrong here then

Reply:
Nah it is additional comment :)